ols/user_data_collection/data_collector.py

```diff
@@ -0,0 +1,313 @@
+    """
+    files = []
+
+    files += list(pathlib.Path(location).glob("feedback/*.json"))
+    files += list(pathlib.Path(location).glob("transcripts/*/*/*.json"))
+
+    return files
+
+
+def package_files_into_tarball(
+    file_paths: list[pathlib.Path], path_to_strip: str
+) -> io.BytesIO:
+    """Package specified directory into a tarball.
+
+    Args:
+        file_paths: List of paths to the files to be packaged.
+        path_to_strip: Path to be stripped from the file paths (not
+            included in the archive).
+
+    Returns:
+        BytesIO object representing the tarball.
+    """
+    tarball_io = io.BytesIO()
+
+    with tarfile.open(fileobj=tarball_io, mode="w:gz") as tar:
+        # arcname parameter is set to a stripped path to avoid including
+        # the full path of the root dir
+        for file_path in file_paths:
+            tar.add(file_path, arcname=file_path.as_posix().replace(path_to_strip, ""))
+
+        # add magic file for identification of our archive on the CCX side
+        empty_file = tarfile.TarInfo("openshift_lightspeed.json")
+        empty_file.size = 0
+        tar.addfile(empty_file)
+
+    tarball_io.seek(0)
+
+    return tarball_io
+
+
+def exponential_backoff_decorator(max_retries: int, base_delay: int) -> Callable:
+    """Exponential backoff decorator."""
+
+    def decorator(func: Callable) -> Callable:
+        def wrapper(*args: Any, **kwargs: Any) -> None:
+            retries = 0
+            while retries < max_retries:
+                try:
+                    func(*args, **kwargs)
+                    return
+                except Exception as e:
+                    logger.error(f"attempt {retries + 1} failed with error: {e}")
+                    retries += 1
+                    delay = base_delay * 2**retries
+                    logger.info(f"retrying in {delay} seconds...")
+                    time.sleep(delay)
+            logger.error("max retries reached, operation failed.")
+
+        return wrapper
+
+    return decorator
+
+
+@exponential_backoff_decorator(
+    max_retries=INGRESS_MAX_RETRIES, base_delay=INGRESS_BASE_DELAY
+)
+def upload_data_to_ingress(tarball: io.BytesIO) -> requests.Response:
+    """Upload the tarball to a Ingress.
+
+    Args:
+        tarball: BytesIO object representing the tarball to be uploaded.
+
+    Returns:
+        Response object from the Ingress.
+    """
+    logger.info("sending collected data")
+    payload = {
+        "file": (
+            "ols.tgz",
+            tarball.read(),
+            "application/vnd.redhat.openshift.periodic+tar",
+        ),
+    }
+
+    if CP_OFFLINE_TOKEN:
+        logger.debug("using CP offline token to generate refresh token")
+        token = access_token_from_offline_token(CP_OFFLINE_TOKEN)
+        cluster_id = "unknown"
+    else:
+        logger.debug("using cluster pull secret to authenticate")
+        token = get_cloud_openshift_pull_secret()
+        cluster_id = get_cluster_id()
```

Comment:
error handling for both of these in the event we can't get the pull secret or cluster id